lib/ansible/modules/cloud/amazon/aws_sgw_facts.py

```diff
@@ -0,0 +1,337 @@
+      description: "Storage gateway local disks"
+      returned: always
+      type: complex
+      contains:
+        disk_allocation_type:
+          description: "Disk allocation type"
+          returned: always
+          type: string
+          sample: "CACHE STORAGE"
+        disk_id:
+          description: "Disk ID on the system"
+          returned: always
+          type: string
+          sample: "pci-0000:00:1f.0"
+        disk_node:
+          description: "Disk parent block device"
+          returned: always
+          type: string
+          sample: "/dev/sdb"
+        disk_path:
+          description: "Disk path used for the cache"
+          returned: always
+          type: string
+          sample: "/dev/nvme1n1"
+        disk_size_in_bytes:
+          description: "Disk size in bytes"
+          returned: always
+          type: integer
+          sample: 107374182400
+        disk_status:
+          description: "Disk status"
+          returned: always
+          type: string
+          sample: "present"
+'''
+
+EXAMPLES = '''
+# Note: These examples do not set authentication details, see the AWS Guide for details.
+
+- name: "Get AWS storage gateway facts"
+  aws_sgw_facts:
+
+- name: "Get AWS storage gateway facts for region eu-west-3"
+  aws_sgw_facts:
+    region: eu-west-3
+'''
+
+from ansible.module_utils.aws.core import AnsibleAWSModule
+from ansible.module_utils.ec2 import (ec2_argument_spec, boto3_conn, HAS_BOTO3, get_aws_connection_info,
+                                      compare_aws_tags, camel_dict_to_snake_dict)
+
+try:
+    import botocore
+    from botocore.exceptions import ClientError
+except ImportError:
+    pass  # caught by imported HAS_BOTO3
+
+
+class SGWFactsManager(object):
+    def __init__(self, connection, module):
+        self.connection = connection
+        self.module = module
+        self.name = self.module.params.get('name')
+
+    def fetch(self):
+        gateways = self.list_gateways()
+        for gateway in gateways:
+            self.list_local_disks(gateway)
+            # File share gateway
+            if gateway["gateway_type"] == "FILE_S3":
+                self.list_gateway_file_shares(gateway)
+            # Volume tape gateway
+            elif gateway["gateway_type"] == "VTL":
+                self.list_gateway_vtl(gateway)
+            # iSCSI gateway
+            elif gateway["gateway_type"] in ["CACHED", "STORED"]:
+                self.list_gateway_volumes(gateway)
+
+        self.module.exit_json(gateways=gateways)
+
+    """
+    Reads the storage gateway object API response.
+    """
+    @staticmethod
+    def _read_gatewaylist_reponse(gateway_list, aws_response):
+        for gw in aws_response["Gateways"]:
+            gateway_list.append(camel_dict_to_snake_dict(gw))
+        return aws_response["Marker"] if "Marker" in aws_response else None
+
+    """
+    List all storage gateways for the AWS endpoint.
+    """
+    def list_gateways(self):
```

Comment:
As Will recommends, paginators will simplify this code very much. There's a hidden gem in paginators called `search` that will also let you pull out all of one key and merge those lists: 

```
pg = self.connection.get_paginator('list_gateways')
return list(pg.paginate().search('Gateways[*]'))

# or, if the result will only be used in a for loop
return pg.paginate().search('Gateways[*]')
# will return an iterator that will build results on the fly
```

This would grab the paginated list of all gateways and merge them all, or if there were no gateways it would return an empty list. 

Example (cloudformation module, but all paginators have this) https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/cloudformation.py#L294

Reply:
I just changed the code to use paginator where needed